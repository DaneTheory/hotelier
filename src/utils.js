import isElectron from 'is-electron';

export default class Utils {
  static openExternalLink = (url) => {
    if (isElectron()) {
      window.shell.openExternal(url);
    } else {
      window.open(url);
    }
  }
}
