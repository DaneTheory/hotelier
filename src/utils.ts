import isElectron from 'is-electron';

export default class Utils {
  static isElectron = () => {
    return isElectron();
  }

  static openExternalLink = (url: string) => {
    if (Utils.isElectron()) {
      window.shell.openExternal(url);
    } else {
      window.open(url);
    }
  }
}
