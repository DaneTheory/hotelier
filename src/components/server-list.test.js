import React from 'react';
import ServerList from './server-list';
import utils from '../utils';
import { shallow } from 'enzyme';
import hotelApi, { STOPPED, RUNNING } from '../api';

describe('ServerList', () => {
  const servers = [
    { id: 'test-server-1', status: 'running' },
    { id: 'test-server-2', status: 'stopped' },
  ];
  let updateServerStatusFn;
  let loadServersFn;

  beforeEach(() => {
    updateServerStatusFn = jest.fn();
    loadServersFn = jest.fn();
  });

  function create() {
    return shallow(<ServerList updateServerStatus={updateServerStatusFn} loadServers={loadServersFn} servers={servers} />);
  }

  it('matches the snapshot', () => {
    const rendered = create();
    expect(rendered).toMatchSnapshot();
  });

  describe('#startServer', () => {
    let instance;

    beforeEach(() => {
      instance = create().instance();
      hotelApi.startServer = jest.fn().mockReturnValue(Promise.resolve());
      instance.startServer(servers[0].id);
    });

    it('calls the api', () => {
      expect(hotelApi.startServer).toHaveBeenCalledWith('test-server-1');
    });

    it('updates the state', () => {
      expect(updateServerStatusFn).toHaveBeenCalledWith('test-server-1', RUNNING);
    });
  });

  describe('#stopServer', () => {
    let instance;

    beforeEach(() => {
      instance = create().instance();
      hotelApi.stopServer = jest.fn().mockReturnValue(Promise.resolve());
      instance.stopServer(servers[0].id);
    });

    it('calls the api', () => {
      expect(hotelApi.stopServer).toHaveBeenCalledWith('test-server-1');
    });

    it('updates the state', () => {
      expect(updateServerStatusFn).toHaveBeenCalledWith('test-server-1', STOPPED);
    });
  });

  describe('#toggleServer', () => {
    let instance;

    beforeEach(() => {
      instance = create().instance();
      hotelApi.startServer = jest.fn().mockReturnValue(Promise.resolve());
      hotelApi.stopServer = jest.fn().mockReturnValue(Promise.resolve());
    });

    it('starts when it is stopped', () => {
      instance.toggleServer(servers[1]);
      expect(hotelApi.startServer).toHaveBeenCalledWith('test-server-2');
    });

    it('stops when it is started', () => {
      instance.toggleServer(servers[0]);
      expect(hotelApi.stopServer).toHaveBeenCalledWith('test-server-1');
    });
  });

  it('can open server and reload servers afterwards', () => {
    jest.useFakeTimers();
    const instance = create().instance();
    jest.spyOn(utils, 'openExternalLink');
    instance.openServer(servers[0]);
    expect(utils.openExternalLink).toHaveBeenCalledWith('http://test-server-1.localhost');
    loadServersFn.mockClear();
    jest.runAllTimers();
    expect(loadServersFn).toHaveBeenCalled();
  });
});
