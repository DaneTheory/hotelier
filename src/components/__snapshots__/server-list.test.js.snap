// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerList matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ServerList
    loadServers={[Function]}
    servers={
        Array [
            Object {
              "id": "test-server-1",
              "status": "running",
            },
            Object {
              "id": "test-server-2",
              "status": "stopped",
            },
          ]
    }
    updateServerStatus={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <table>
        <thead>
                <tr>
                        <th>
                                Server
                        </th>
                        <th>
                                Action
                        </th>
                </tr>
        </thead>
        <tbody>
                <ServerItem
                        openServer={[Function]}
                        server={
                                Object {
                                        "id": "test-server-1",
                                        "status": "running",
                                      }
                        }
                        toggleServer={[Function]}
                />
                <ServerItem
                        openServer={[Function]}
                        server={
                                Object {
                                        "id": "test-server-2",
                                        "status": "stopped",
                                      }
                        }
                        toggleServer={[Function]}
                />
        </tbody>
</table>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <thead>
            <tr>
                        <th>
                                    Server
                        </th>
                        <th>
                                    Action
                        </th>
            </tr>
</thead>,
          <tbody>
            <ServerItem
                        openServer={[Function]}
                        server={
                                    Object {
                                                "id": "test-server-1",
                                                "status": "running",
                                              }
                        }
                        toggleServer={[Function]}
            />
            <ServerItem
                        openServer={[Function]}
                        server={
                                    Object {
                                                "id": "test-server-2",
                                                "status": "stopped",
                                              }
                        }
                        toggleServer={[Function]}
            />
</tbody>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <tr>
              <th>
                            Server
              </th>
              <th>
                            Action
              </th>
</tr>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <th>
                  Server
</th>,
                <th>
                  Action
</th>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Server",
                },
                "ref": null,
                "rendered": "Server",
                "type": "th",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Action",
                },
                "ref": null,
                "rendered": "Action",
                "type": "th",
              },
            ],
            "type": "tr",
          },
          "type": "thead",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <ServerItem
                openServer={[Function]}
                server={
                                Object {
                                                "id": "test-server-1",
                                                "status": "running",
                                              }
                }
                toggleServer={[Function]}
/>,
              <ServerItem
                openServer={[Function]}
                server={
                                Object {
                                                "id": "test-server-2",
                                                "status": "stopped",
                                              }
                }
                toggleServer={[Function]}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "test-server-1",
              "nodeType": "class",
              "props": Object {
                "openServer": [Function],
                "server": Object {
                  "id": "test-server-1",
                  "status": "running",
                },
                "toggleServer": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "test-server-2",
              "nodeType": "class",
              "props": Object {
                "openServer": [Function],
                "server": Object {
                  "id": "test-server-2",
                  "status": "stopped",
                },
                "toggleServer": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "tbody",
        },
      ],
      "type": "table",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <table>
          <thead>
                    <tr>
                              <th>
                                        Server
                              </th>
                              <th>
                                        Action
                              </th>
                    </tr>
          </thead>
          <tbody>
                    <ServerItem
                              openServer={[Function]}
                              server={
                                        Object {
                                                  "id": "test-server-1",
                                                  "status": "running",
                                                }
                              }
                              toggleServer={[Function]}
                    />
                    <ServerItem
                              openServer={[Function]}
                              server={
                                        Object {
                                                  "id": "test-server-2",
                                                  "status": "stopped",
                                                }
                              }
                              toggleServer={[Function]}
                    />
          </tbody>
</table>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <thead>
              <tr>
                            <th>
                                          Server
                            </th>
                            <th>
                                          Action
                            </th>
              </tr>
</thead>,
            <tbody>
              <ServerItem
                            openServer={[Function]}
                            server={
                                          Object {
                                                        "id": "test-server-1",
                                                        "status": "running",
                                                      }
                            }
                            toggleServer={[Function]}
              />
              <ServerItem
                            openServer={[Function]}
                            server={
                                          Object {
                                                        "id": "test-server-2",
                                                        "status": "stopped",
                                                      }
                            }
                            toggleServer={[Function]}
              />
</tbody>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tr>
                <th>
                                Server
                </th>
                <th>
                                Action
                </th>
</tr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <th>
                    Server
</th>,
                  <th>
                    Action
</th>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Server",
                  },
                  "ref": null,
                  "rendered": "Server",
                  "type": "th",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Action",
                  },
                  "ref": null,
                  "rendered": "Action",
                  "type": "th",
                },
              ],
              "type": "tr",
            },
            "type": "thead",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <ServerItem
                  openServer={[Function]}
                  server={
                                    Object {
                                                      "id": "test-server-1",
                                                      "status": "running",
                                                    }
                  }
                  toggleServer={[Function]}
/>,
                <ServerItem
                  openServer={[Function]}
                  server={
                                    Object {
                                                      "id": "test-server-2",
                                                      "status": "stopped",
                                                    }
                  }
                  toggleServer={[Function]}
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "test-server-1",
                "nodeType": "class",
                "props": Object {
                  "openServer": [Function],
                  "server": Object {
                    "id": "test-server-1",
                    "status": "running",
                  },
                  "toggleServer": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "test-server-2",
                "nodeType": "class",
                "props": Object {
                  "openServer": [Function],
                  "server": Object {
                    "id": "test-server-2",
                    "status": "stopped",
                  },
                  "toggleServer": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "tbody",
          },
        ],
        "type": "table",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
